### PR TITLE
Word mapping support

### PR DESCRIPTION
**Pull request by irvanjitsingh**
Adding support to use a full word-mapping for gurmukhi transliteration.

Use-cases:
- languages that are difficult to transliterate using a per-letter mapping
- ability to handle every edge case by using a completely custom word mapping

Word mapping transliteration is enabled by passing in an optional Map object. Use the wordMap object to generate a single instance of the map and pass it wherever the translit function is called

If a given word is not found in the map, then the specified transliteration module is used.

Possible speed and memory improvements: parsing the CSV and generating the map.

### CODE REVIEW

**Review by tarunsingh5 — APPROVED** [review]
looked at first 3 ish files and lgtm

**Comment by ManjotS** [comment]
where is the actual map?

**Comment by irvanjitsingh** [comment]
> where is the actual map?

It gets generated once on runtime from a csv provided by the user using the wordMap function. I took out the sample csv to reduce module bloating.

File: src/translit.js [hunk 1]
```diff
@@ -9,21 +9,43 @@ languages.ipa = require('./translit_modules/ipa');
...
-function translit(gurmukhi, language = 'english') {
+function translit(gurmukhi, language = 'english', map = null) {
   if (language === 'all') {
     return Object.keys(languages).reduce((out, x) => {
       /* eslint no-param-reassign: 0 */
       out[x] = languages[x](gurmukhi);
       return out;
     }, {});
   }
+  function getMappedWord(word) {
```
Comment (ManjotS) [inline src/translit.js]: why did you do a function in a function?

File: src/wordMap.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/* load modules */
```
Comment (ManjotS) [inline src/wordMap.js]: can we rename to src/translitWordMap.js or move it to translit modules dir?

**Comment by irvanjitsingh** [comment]
@ManjotS @tsingh777 please review when you get a chance

**Review by ManjotS — APPROVED** [review]

File: src/translit_modules/english_map_example.csv [hunk 3]
```diff
@@ -0,0 +1,6 @@
+﻿gurmukhi,english
```
Comment (tsingh777) [inline src/translit_modules/english_map_example.csv:1]: This file should go into the `__tests__` directory.

File: src/translitWordMap.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Returns a map of words to be used in a word-by-word transliteration
+ *
+ * @since 1.0.0
+ * @param {string} filePath The path to the csv file representing the word map
+ * @returns {Map} Returns a Map object representation of the input csv file
+ */
+
+function translitWordMap(filePath) {
```
Comment (tsingh777) [inline src/translitWordMap.js:13]: better name (for the file as well) would be `readCsvWordMap`. is this just for the test or do we expect user's to call this (not likely IMO).

File: src/translit.js [hunk 5]
```diff
@@ -10,22 +10,41 @@ languages.shahmukhi = require('./translit_modules/shahmukhi');
...
       return out;
     }, {});
   }
+
+  if (map != null) {
+    return gurmukhi.split(' ').map((word) => {
+      // remove embedded vishraams before checking map
+      const vishraamRegex = /[.,;]$/;
+      const vishraam = word.match(vishraamRegex);
+      word = word.replace(vishraamRegex, '');
```
Comment (tsingh777) [inline src/translit.js:36]: this works but it **not** good. https://github.com/airbnb/javascript#functions--reassign-params. I will add a linter to enforce best practices.

File: src/translitWordMap.js (hunk 4 above)
Comment (tsingh777) [reply src/translitWordMap.js:13]: if its just for the test then I'd remove.

**Review by tsingh777 — COMMENTED** [review]
@irvanjitsingh please follow up when you get a chance.

File: src/translitWordMap.js (hunk 4 above)
Comment (irvanjitsingh) [reply src/translitWordMap.js:13]: Its a user callable function by design; the idea is to provide a utility for the user to generate a map from a simple csv file, and also abstract this functionality. Performance-wise, the best practice would be the user generates a map once in their script and then passes it in wherever the translit function is called.

Comment (irvanjitsingh) [reply src/translitWordMap.js:13]: I'm not sure if its a good idea to except the user to have their own Javascript Map object ready to use with anvaad.

**Pull request merged**